File: scr/P119.py
import time 

def timer(func):
    def wrapper(*args, **kwargs):
        before = time.time()
        r = func(*args, **kwargs)
        name = str(func).split(' ')[1]
        print( f'Resuldo of {name} : {r} (execution time {time.time()-before}s)' ) 
        return r

    return wrapper


@timer
def P119(N:int=30) -> int: 
    # Optimized approach
    numbers_with_property = []

    def sum_of_digits(n):
        """Return the sum of the digits of n."""
        return sum(int(digit) for digit in str(n))

    # Consider sums from 2 (smallest 2-digit number is 10) to 9*7 (largest 7-digit number is 9999999)
    for digit_sum in range(2, 300 + 1):
        i = 1

        while True:
            num = digit_sum ** i
            if sum_of_digits(num) == digit_sum and num > 10:
                numbers_with_property.append(num)
            # If number of digits in num > digit_sum, break the loop
            if len(str(num)) > digit_sum:
                break
            i += 1

    # Sort the numbers and get the 30th number
    numbers_with_property = sorted(numbers_with_property)
    return  numbers_with_property[:30]

P119() #  248155780267521

File: scr/P119.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <ctime>
#include <string>
#include <cmath>

using namespace std;

// Function to get the sum of digits of a number
int sum_of_digits(long long n) {
    int sum = 0;
    while (n) {
        sum += n % 10;
        n /= 10;
    }
    return sum;
}

// Timer decorator equivalent
template<typename Func>
auto timer(Func func, int N) {
    clock_t start = clock();
    auto result = func(N);
    double duration = (clock() - start) / (double)CLOCKS_PER_SEC;
    cout << "Result: " << result << " (execution time " << duration << "s)" << endl;
    return result;
}

long long P119(int N = 30) {
    vector<long long> numbers_with_property;


    // Max value for long long
    const long long MAX_LL = numeric_limits<long long>::max();

    // Consider sums from 2 to 300 (chosen based on Python code)
    for (int digit_sum = 2; digit_sum <= 500; ++digit_sum) {
        int i = 1;

        while (true) {
            // Check for potential overflow
            if (digit_sum > pow(MAX_LL, 1.0/i)) {
                break;
            }


            long long num = pow(digit_sum, i);
            cout << "Result: " << num << " (ds " << digit_sum << " (numL "  << to_string(num).length() << " (execution time " << endl;
            if (sum_of_digits(num) == digit_sum && num > 10) {
                numbers_with_property.push_back(num);
            }

            // If number of digits in num > digit_sum, break the loop
            if (to_string(num).length() > digit_sum) {
                break;
            }

            i++;
        }
    }

    // Sort the numbers
    sort(numbers_with_property.begin(), numbers_with_property.end());

    // Return the Nth number
    return numbers_with_property[N-1];
}

int main() {
    // Call the function with timer
    timer(P119, 30);

    return 0;
}
